subsys/mgmt/mcumgr/lib/smp/src/smp.c

```diff
@@ -55,57 +60,34 @@ smp_read_hdr(struct smp_streamer *streamer, struct mgmt_hdr *dst_hdr)
 	return 0;
 }
 
-static int
+static inline int
 smp_write_hdr(struct smp_streamer *streamer, const struct mgmt_hdr *src_hdr)
 {
 	int rc;
 
-	rc = mgmt_streamer_write_at(&streamer->mgmt_stmr, 0, src_hdr, sizeof(*src_hdr));
+	rc = mgmt_streamer_write_hdr(&streamer->mgmt_stmr, src_hdr);
 	return mgmt_err_from_cbor(rc);
 }
 
 static int
-smp_build_err_rsp(struct smp_streamer *streamer,
-				  const struct mgmt_hdr *req_hdr,
-				  int status)
+smp_build_err_rsp(struct smp_streamer *streamer, const struct mgmt_hdr *req_hdr, int status,
+		  const char *rc_rsn)
 {
-	struct CborEncoder map;
-	struct mgmt_ctxt cbuf;
 	struct mgmt_hdr rsp_hdr;
-	int rc;
-
-	rc = mgmt_ctxt_init(&cbuf, &streamer->mgmt_stmr);
-	if (rc != 0) {
-		return rc;
-	}
-
-	smp_init_rsp_hdr(req_hdr, &rsp_hdr);
-	rc = smp_write_hdr(streamer, &rsp_hdr);
-	if (rc != 0) {
-		return rc;
-	}
-
-	rc = cbor_encoder_create_map(&cbuf.encoder, &map, CborIndefiniteLength);
-	if (rc != 0) {
-		return rc;
-	}
-
-	rc = mgmt_write_rsp_status(&cbuf, status);
-	if (rc != 0) {
-		return rc;
-	}
-
-	rc = cbor_encoder_close_container(&cbuf.encoder, &map);
-	if (rc != 0) {
-		return rc;
+	struct cbor_nb_writer *nbw = (struct cbor_nb_writer *)streamer->mgmt_stmr.writer;
+	zcbor_state_t *zsp = &nbw->zs;
+
+	zcbor_map_start_encode(zsp, 1);
+	zcbor_tstr_put_lit(zsp, "rc");
+	zcbor_int32_put(zsp, status);
```

Comment:
Ok, makes sense, I'd recommend either a comment or an assert statement to this effect. BTW zcbor recently gained an error variable in the state struct.

Reply:
Would add in the check, it's possible that an application has a memory corruption issue and if this is triggered, it can be apparent at the time that something is very wrong